server/schedule/operator/step.go

```diff
@@ -437,3 +449,323 @@ func (al AddLightLearner) Influence(opInfluence OpInfluence, region *core.Region
+		}
+		return peer.Id == df.PeerID
+	}
+	return false
+}
+
+// CheckSafety checks if the step meets the safety properties.
+func (df DemoteFollower) CheckSafety(region *core.RegionInfo) error {
+	peer := region.GetStorePeer(df.ToStore)
+	if peer == nil || peer.Id != df.PeerID {
+		return errors.New("peer does not exist")
+	}
+	if peer.Id == region.GetLeader().GetId() {
+		return errors.New("peer is a leader, can not demote")
+	}
+	return nil
+}
+
+// Influence calculates the store difference that current step makes.
+func (df DemoteFollower) Influence(opInfluence OpInfluence, region *core.RegionInfo) {}
+
+// DemoteVoter is very similar to DemoteFollower. But it allows Demote Leader.
+// Note: It is not an OpStep, only a sub step in ChangePeerV2Enter.
+type DemoteVoter struct {
+	ToStore, PeerID uint64
+}
+
+func (dv DemoteVoter) String() string {
+	return fmt.Sprintf("demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+}
+
+// ConfVerChanged returns how much the conf version will increase due to this step.
+func (dv DemoteVoter) ConfVerChanged(region *core.RegionInfo) bool {
+	peer := region.GetStoreLearner(dv.ToStore)
+	return peer.GetId() == dv.PeerID
+}
+
+// IsFinish checks if current step is finished.
+func (dv DemoteVoter) IsFinish(region *core.RegionInfo) bool {
+	if peer := region.GetStoreLearner(dv.ToStore); peer != nil {
+		if peer.Id != dv.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", dv.String()), zap.Uint64("obtain-learner", peer.Id))
+		}
+		return peer.Id == dv.PeerID
+	}
+	return false
+}
+
+// ChangePeerV2Enter is an OpStep that uses joint consensus to request all PromoteLearner and DemoteVoter.
+type ChangePeerV2Enter struct {
+	PromoteLearners []PromoteLearner
+	DemoteVoters    []DemoteVoter
+}
+
+func (cpe ChangePeerV2Enter) String() string {
+	b := &strings.Builder{}
+	_, _ = b.WriteString("use joint consensus")
+	for _, pl := range cpe.PromoteLearners {
+		_, _ = fmt.Fprintf(b, ", promote learner peer %v on store %v to voter", pl.PeerID, pl.ToStore)
+	}
+	for _, dv := range cpe.DemoteVoters {
+		_, _ = fmt.Fprintf(b, ", demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+	}
+	return b.String()
+}
+
+// ConfVerChanged returns how much the conf version will increase due to this step.
+func (cpe ChangePeerV2Enter) ConfVerChanged(region *core.RegionInfo) uint64 {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer == nil || peer.Id != pl.PeerID ||
+			(peer.Role != metapb.PeerRole_IncomingVoter && peer.Role != metapb.PeerRole_Voter) {
+			return 0
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && (peer.Id != dv.PeerID || (peer.Role != metapb.PeerRole_DemotingVoter && peer.Role != metapb.PeerRole_Learner)) {
+			return 0
+		}
+	}
+	return uint64(len(cpe.PromoteLearners) + len(cpe.DemoteVoters))
```

Comment:
Does it mean the config version only updates when all changes are done?

Reply:
Yes, TiKV side will change atomically.